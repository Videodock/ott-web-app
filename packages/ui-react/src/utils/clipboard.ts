export const copyToClipboard = (value: string): void => {
  const inputElement = document.createElement('input');
  inputElement.style.zIndex = '-10';
  inputElement.style.position = 'absolute';
  inputElement.style.left = '-1000';
  inputElement.value = value;
  document.body.appendChild(inputElement);
  inputElement.select();
  document.execCommand('copy');
  inputElement.blur();
  document.body.removeChild(inputElement);
};
