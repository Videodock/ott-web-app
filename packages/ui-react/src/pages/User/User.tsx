import { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Navigate, Route, Routes, useLocation, useNavigate } from 'react-router-dom';
import { shallow } from '@jwp/ott-common/src/utils/compare';
import { getModule } from '@jwp/ott-common/src/modules/container';
import { useAccountStore } from '@jwp/ott-common/src/stores/AccountStore';
import { useConfigStore } from '@jwp/ott-common/src/stores/ConfigStore';
import FavoritesController from '@jwp/ott-common/src/stores/FavoritesController';
import AccountController from '@jwp/ott-common/src/stores/AccountController';
import CheckoutController from '@jwp/ott-common/src/stores/CheckoutController';
import { useProfileStore } from '@jwp/ott-common/src/stores/ProfileStore';
import { ACCESS_MODEL, PersonalShelf } from '@jwp/ott-common/src/constants';
import useBreakpoint, { Breakpoint } from '@jwp/ott-ui-react/src/hooks/useBreakpoint';
import { useProfiles } from '@jwp/ott-hooks-react/src/useProfiles';

import AccountCircle from '../../icons/AccountCircle';
import BalanceWallet from '../../icons/BalanceWallet';
import Exit from '../../icons/Exit';
import Favorite from '../../icons/Favorite';
import AccountComponent from '../../components/Account/Account';
import Button from '../../components/Button/Button';
import ConfirmationDialog from '../../components/ConfirmationDialog/ConfirmationDialog';
import Favorites from '../../components/Favorites/Favorites';
import LoadingOverlay from '../../components/LoadingOverlay/LoadingOverlay';
import PaymentContainer from '../../containers/PaymentContainer/PaymentContainer';
import PlaylistContainer from '../../containers/PlaylistContainer/PlaylistContainer';
import EditProfile from '../../containers/Profiles/EditProfile';

import styles from './User.module.scss';

const User = (): JSX.Element => {
  const favoritesController = getModule(FavoritesController);
  const accountController = getModule(AccountController);
  const checkoutController = getModule(CheckoutController);

  const { accessModel, favoritesList } = useConfigStore(
    (s) => ({
      accessModel: s.accessModel,
      favoritesList: s.config.features?.favoritesList,
    }),
    shallow,
  );
  const navigate = useNavigate();
  const { t } = useTranslation('user');
  const breakpoint = useBreakpoint();
  const [clearFavoritesOpen, setClearFavoritesOpen] = useState(false);
  const location = useLocation();

  const isLargeScreen = breakpoint > Breakpoint.md;
  const { user: customer, subscription, loading } = useAccountStore();
  const { canUpdateEmail, canRenewSubscription } = accountController.getFeatures();
  const { profile } = useProfileStore();

  const { profilesEnabled } = useProfiles();

  const profileAndFavoritesPage = location.pathname?.includes('my-profile') || location.pathname.includes('favorites');

  const onLogout = useCallback(async () => {
    // Empty customer on a user page leads to navigate (code bellow), so we don't repeat it here
    await accountController.logout();
  }, [accountController]);

  useEffect(() => {
    if (accessModel !== ACCESS_MODEL.AVOD && canRenewSubscription) {
      checkoutController.getSubscriptionSwitches();
    }
  }, [accessModel, checkoutController, canRenewSubscription]);

  useEffect(() => {
    if (!loading && !customer) {
      navigate('/', { replace: true });
    }
  }, [navigate, customer, loading]);

  if (!customer) {
    return (
      <div className={styles.user}>
        <LoadingOverlay inline />
      </div>
    );
  }

  return (
    <div className={styles.user}>
      {isLargeScreen && (
        <div className={styles.leftColumn}>
          <div className={styles.panel}>
            <ul>
              {accessModel === 'SVOD' && profilesEnabled && profileAndFavoritesPage && (
                <li>
                  <Button
                    to={`my-profile/${profile?.id}`}
                    label={profile?.name ?? t('nav.profile')}
                    variant="text"
                    startIcon={<img className={styles.profileIcon} src={profile?.avatar_url} alt={profile?.name} />}
                    className={styles.button}
                  />
                </li>
              )}
              {(!profilesEnabled || !profileAndFavoritesPage) && (
                <li>
                  <Button to="my-account" label={t('nav.account')} variant="text" startIcon={<AccountCircle />} className={styles.button} />
                </li>
              )}
              {favoritesList && (!profilesEnabled || profileAndFavoritesPage) && (
                <li>
                  <Button to="favorites" label={t('nav.favorites')} variant="text" startIcon={<Favorite />} className={styles.button} />
                </li>
              )}

              {accessModel !== ACCESS_MODEL.AVOD && (!profilesEnabled || !profileAndFavoritesPage) && (
                <li>
                  <Button to="payments" label={t('nav.payments')} variant="text" startIcon={<BalanceWallet />} className={styles.button} />
                </li>
              )}

              {(!profilesEnabled || !profileAndFavoritesPage) && (
                <li className={styles.logoutLi}>
                  <Button onClick={onLogout} label={t('nav.logout')} variant="text" startIcon={<Exit />} className={styles.button} />
                </li>
              )}
            </ul>
          </div>
        </div>
      )}
      <div className={styles.mainColumn}>
        <Routes>
          <Route
            path="my-account"
            element={<AccountComponent panelClassName={styles.panel} panelHeaderClassName={styles.panelHeader} canUpdateEmail={canUpdateEmail} />}
          />
          {profilesEnabled && <Route path="my-profile/:id" element={<EditProfile contained />} />}
          {favoritesList && (
            <Route
              path="favorites"
              element={
                <>
                  <PlaylistContainer type={PersonalShelf.Favorites} showEmpty>
                    {({ playlist, error, isLoading }) => (
                      <Favorites
                        playlist={playlist}
                        error={error}
                        isLoading={isLoading}
                        onClearFavoritesClick={() => setClearFavoritesOpen(true)}
                        accessModel={accessModel}
                        hasSubscription={!!subscription}
                      />
                    )}
                  </PlaylistContainer>
                  <ConfirmationDialog
                    open={clearFavoritesOpen}
                    title={t('favorites.clear_favorites_title')}
                    body={t('favorites.clear_favorites_body')}
                    onConfirm={async () => {
                      await favoritesController.clear();
                      setClearFavoritesOpen(false);
                    }}
                    onClose={() => setClearFavoritesOpen(false)}
                  />
                </>
              }
            />
          )}
          <Route path="payments" element={accessModel !== ACCESS_MODEL.AVOD ? <PaymentContainer /> : <Navigate to="my-account" />} />
          <Route path="*" element={<Navigate to="my-account" />} />
        </Routes>
      </div>
    </div>
  );
};

export default User;
