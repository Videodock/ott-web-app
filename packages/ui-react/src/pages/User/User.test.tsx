import { act } from '@testing-library/react';
import type { Playlist, PlaylistItem } from '@jwp/ott-common/types/playlist';
import type { Config } from '@jwp/ott-common/types/config';
import type { PaymentDetail, Subscription, Transaction } from '@jwp/ott-common/types/subscription';
import { useAccountStore } from '@jwp/ott-common/src/stores/AccountStore';
import { useConfigStore } from '@jwp/ott-common/src/stores/ConfigStore';
import { useFavoritesStore } from '@jwp/ott-common/src/stores/FavoritesStore';
import AccountController from '@jwp/ott-common/src/stores/AccountController';
import { mockService } from '@jwp/ott-common/test/mockService';
import ApiService from '@jwp/ott-common/src/services/ApiService';
import FavoritesController from '@jwp/ott-common/src/stores/FavoritesController';
import CheckoutController from '@jwp/ott-common/src/stores/CheckoutController';
import ProfileController from '@jwp/ott-common/src/stores/ProfileController';
import { ACCESS_MODEL, DEFAULT_FEATURES } from '@jwp/ott-common/src/constants';
import { Route, Routes } from 'react-router-dom';
import React from 'react';

import { mockWindowLocation, renderWithRouter } from '../../../test/utils';

import User from './User';

const data = {
  loading: false,
  user: {
    id: 'user1',
    email: 'user@jwplayer.com',
    country: 'us',
    firstName: 'User',
    lastName: 'Person',
    regDate: 'Jan 1, 2000',
    lastUserIp: '192.0.0.1',
  },
  subscription: {
    subscriptionId: 90210,
    offerId: 'offer1111',
    expiresAt: 1767994,
    status: 'active',
    nextPaymentPrice: 199.99,
    nextPaymentCurrency: 'USD',
    paymentGateway: '',
    paymentMethod: 'Paypal',
    offerTitle: 'Your Cool Subscription',
    period: 'year',
    totalPrice: 99.95,
  } as Subscription,
  transactions: [
    {
      transactionId: '11232',
      offerId: 'offer1111',
      paymentMethod: 'Ideal',
      offerTitle: 'Your Cool Subscription',
      transactionPriceInclTax: 12.88,
      transactionCurrency: 'EUR',
      customerCountry: 'nl',
      transactionDate: '1657994000',
    } as unknown as Transaction,
    {
      transactionId: '11234',
      offerId: 'offer13',
      paymentMethod: 'Cash',
      offerTitle: 'Your Cool Subscription',
      transactionPriceInclTax: 12.88,
      transactionCurrency: 'EUR',
      customerCountry: 'uk',
      transactionDate: '1667990000',
    } as unknown as Transaction,
  ],
  activePayment: {
    id: 887765,
    paymentMethodSpecificParams: {
      lastCardFourDigits: '9888',
      cardExpirationDate: '03/30',
    },
  } as unknown as PaymentDetail,
};

describe('User Component tests', () => {
  beforeEach(() => {
    mockService(ApiService, {});
    mockService(AccountController, {
      logout: vi.fn(),
      getFeatures: vi.fn(() => ({
        ...DEFAULT_FEATURES,
        canUpdateEmail: false,
        canRenewSubscription: false,
      })),
    });
    mockService(FavoritesController, { clear: vi.fn() });
    mockService(CheckoutController, { getSubscriptionSwitches: vi.fn(), getSubscriptionOfferIds: vi.fn().mockReturnValue([]) });
    mockService(ProfileController, { listProfiles: vi.fn(), isEnabled: vi.fn().mockReturnValue(false) });

    useConfigStore.setState({
      accessModel: ACCESS_MODEL.SVOD,
    });
  });

  test('Account Page', () => {
    act(() => {
      useAccountStore.setState(data);
      mockWindowLocation('u/my-account');
    });
    const { container } = renderWithRouter(
      <Routes>
        <Route path="/u/*" element={<User />} />
      </Routes>,
    );

    expect(container).toMatchSnapshot();
  });

  test('Payments Page', () => {
    act(() => {
      useAccountStore.setState(data);
      mockWindowLocation('u/payments');
    });
    const { container } = renderWithRouter(
      <Routes>
        <Route path="/u/*" element={<User />} />
      </Routes>,
    );

    expect(container).toMatchSnapshot();
  });

  test('Favorites Page', () => {
    act(() => {
      useAccountStore.setState(data);
      useConfigStore.setState({
        config: {
          features: {
            favoritesList: 'abcdefgh',
          },
          styling: {},
        } as unknown as Config,
      });
      useFavoritesStore.setState({
        getPlaylist: (): Playlist => {
          return {
            title: 'These are my favorite things',
            playlist: [
              {
                mediaid: 'aaabbbcc',
                title: 'Fav 1',
                duration: 12,
                feedid: 'abcdffff',
              } as unknown as PlaylistItem,
              {
                mediaid: 'aaabbbcd',
                title: 'Big Buck Bunny',
                duration: 6000,
                feedid: 'bbbdddff',
              } as unknown as PlaylistItem,
              {
                mediaid: 'ggaaccvv',
                title: 'My last favorite',
                duration: 659,
                feedid: 'bbbbbbbb',
              } as unknown as PlaylistItem,
            ],
          };
        },
      });
      mockWindowLocation('u/favorites');
    });

    const { container } = renderWithRouter(
      <Routes>
        <Route path="/u/*" element={<User />} />
      </Routes>,
    );

    expect(container).toMatchSnapshot();
  });
});
