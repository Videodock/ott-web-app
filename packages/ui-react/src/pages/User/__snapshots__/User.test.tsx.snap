// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Component tests > Account Page 1`] = `
<div>
  <div
    class="_user_e9c5ca"
  >
    <nav
      class="_leftColumn_e9c5ca"
    >
      <div
        class="_panel_e9c5ca"
      >
        <ul>
          <li>
            <a
              aria-current="page"
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296 _active_f8f296 "
              href="/u/my-account"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12,19.2C9.5,19.2 7.29,17.92 6,16C6.03,14 10,12.9 12,12.9C14,12.9 17.97,14 18,16C16.71,17.92 14.5,19.2 12,19.2M12,5A3,3 0 0,1 15,8A3,3 0 0,1 12,11A3,3 0 0,1 9,8A3,3 0 0,1 12,5M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
                  />
                </svg>
              </div>
              <span>
                nav.account
              </span>
            </a>
          </li>
          <li>
            <a
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              href="/u/payments"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5,3C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V16.72C21.59,16.37 22,15.74 22,15V9C22,8.26 21.59,7.63 21,7.28V5A2,2 0 0,0 19,3H5M5,5H19V7H13A2,2 0 0,0 11,9V15A2,2 0 0,0 13,17H19V19H5V5M13,9H20V15H13V9M16,10.5A1.5,1.5 0 0,0 14.5,12A1.5,1.5 0 0,0 16,13.5A1.5,1.5 0 0,0 17.5,12A1.5,1.5 0 0,0 16,10.5Z"
                  />
                </svg>
              </div>
              <span>
                nav.payments
              </span>
            </a>
          </li>
          <li
            class="_logoutLi_e9c5ca"
          >
            <button
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              type="button"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19,3H5C3.89,3 3,3.89 3,5V9H5V5H19V19H5V15H3V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V5C21,3.89 20.1,3 19,3M10.08,15.58L11.5,17L16.5,12L11.5,7L10.08,8.41L12.67,11H3V13H12.67L10.08,15.58Z"
                  />
                </svg>
              </div>
              <span>
                nav.logout
              </span>
            </button>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="_mainColumn_e9c5ca"
    >
      <h1
        class="hideUntilFocus"
      >
        nav.account
      </h1>
      <section
        aria-labelledby="account_about_you_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="account_about_you_1235"
          >
            account.about_you
          </h2>
        </div>
        <form
          class="_flexBox_1c1c63"
          novalidate=""
        >
          <div
            class="_textField_e16c1b"
          >
            <label
              class="_label_e16c1b"
              for="text-field_1235_firstname"
            >
              account.firstname
            </label>
            <input
              aria-describedby="helper_text_1235_firstname"
              aria-invalid="false"
              autocomplete="given-name"
              class="_input_e16c1b"
              id="text-field_1235_firstname"
              lang="en"
              name="firstName"
              readonly=""
              type="text"
              value="User"
            />
          </div>
          <div
            class="_textField_e16c1b"
          >
            <label
              class="_label_e16c1b"
              for="text-field_1235_lastname"
            >
              account.lastname
            </label>
            <input
              aria-describedby="helper_text_1235_lastname"
              aria-invalid="false"
              autocomplete="family-name"
              class="_input_e16c1b"
              id="text-field_1235_lastname"
              lang="en"
              name="lastName"
              readonly=""
              type="text"
              value="Person"
            />
          </div>
        </form>
        <div
          class="_controls_1c1c63"
        >
          <button
            class="_button_f8f296 _default_f8f296 _outlined_f8f296"
            type="button"
          >
            <span>
              account.edit_information
            </span>
          </button>
        </div>
      </section>
      <section
        aria-labelledby="account_email_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="account_email_1235"
          >
            account.email
          </h2>
        </div>
        <form
          class="_flexBox_1c1c63"
          novalidate=""
        >
          <div
            class="_textField_e16c1b"
          >
            <label
              class="_label_e16c1b"
              for="text-field_1235_email"
            >
              account.email
            </label>
            <input
              aria-describedby="helper_text_1235_email"
              aria-invalid="false"
              autocomplete="email"
              class="_input_e16c1b"
              id="text-field_1235_email"
              name="email"
              readonly=""
              required=""
              type="text"
              value="user@jwplayer.com"
            />
          </div>
        </form>
        <div
          class="_controls_1c1c63"
        />
      </section>
      <section
        aria-labelledby="account_security_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="account_security_1235"
          >
            account.security
          </h2>
        </div>
        <form
          class="_flexBox_1c1c63"
          novalidate=""
        />
        <div
          class="_controls_1c1c63"
        >
          <button
            class="_button_f8f296 _default_f8f296 _outlined_f8f296"
            type="button"
          >
            <span>
              account.edit_password
            </span>
          </button>
        </div>
      </section>
      <section
        aria-labelledby="account_terms_and_tracking_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="account_terms_and_tracking_1235"
          >
            account.terms_and_tracking
          </h2>
        </div>
        <form
          class="_flexBox_1c1c63"
          novalidate=""
        />
        <div
          class="_controls_1c1c63"
        />
      </section>
    </div>
  </div>
</div>
`;

exports[`User Component tests > Favorites Page 1`] = `
<div>
  <div
    class="_user_e9c5ca"
  >
    <nav
      class="_leftColumn_e9c5ca"
    >
      <div
        class="_panel_e9c5ca"
      >
        <ul>
          <li>
            <a
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              href="/u/my-account"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12,19.2C9.5,19.2 7.29,17.92 6,16C6.03,14 10,12.9 12,12.9C14,12.9 17.97,14 18,16C16.71,17.92 14.5,19.2 12,19.2M12,5A3,3 0 0,1 15,8A3,3 0 0,1 12,11A3,3 0 0,1 9,8A3,3 0 0,1 12,5M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
                  />
                </svg>
              </div>
              <span>
                nav.account
              </span>
            </a>
          </li>
          <li>
            <a
              aria-current="page"
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296 _active_f8f296 "
              href="/u/favorites"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                  />
                </svg>
              </div>
              <span>
                nav.favorites
              </span>
            </a>
          </li>
          <li>
            <a
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              href="/u/payments"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5,3C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V16.72C21.59,16.37 22,15.74 22,15V9C22,8.26 21.59,7.63 21,7.28V5A2,2 0 0,0 19,3H5M5,5H19V7H13A2,2 0 0,0 11,9V15A2,2 0 0,0 13,17H19V19H5V5M13,9H20V15H13V9M16,10.5A1.5,1.5 0 0,0 14.5,12A1.5,1.5 0 0,0 16,13.5A1.5,1.5 0 0,0 17.5,12A1.5,1.5 0 0,0 16,10.5Z"
                  />
                </svg>
              </div>
              <span>
                nav.payments
              </span>
            </a>
          </li>
          <li
            class="_logoutLi_e9c5ca"
          >
            <button
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              type="button"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19,3H5C3.89,3 3,3.89 3,5V9H5V5H19V19H5V15H3V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V5C21,3.89 20.1,3 19,3M10.08,15.58L11.5,17L16.5,12L11.5,7L10.08,8.41L12.67,11H3V13H12.67L10.08,15.58Z"
                  />
                </svg>
              </div>
              <span>
                nav.logout
              </span>
            </button>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="_mainColumn_e9c5ca"
    >
      <div>
        <div
          class="_header_546956"
        >
          <h1>
            favorites.title
          </h1>
          <button
            class="_button_f8f296 _default_f8f296 _outlined_f8f296"
            type="button"
          >
            <span>
              favorites.clear
            </span>
          </button>
        </div>
        <div>
          <div
            aria-rowcount="1"
            class="_container_7175d1 _cols-3_7175d1"
            role="grid"
          >
            <div
              aria-rowindex="0"
              class="_row_19b484"
              role="row"
            >
              <div
                aria-colindex="0"
                class="_cell_19b484"
                id="layout_grid_0-0"
                role="gridcell"
                style="width: 33%;"
              >
                <a
                  class="_card_d75732"
                  data-testid="Fav 1"
                  href="/m/aaabbbcc/fav-1?r=abcdffff"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="_titleContainer_d75732"
                  >
                    <h3
                      class="_title_d75732"
                    >
                      Fav 1
                    </h3>
                  </div>
                  <div
                    class="_poster_d75732 _aspect169_d75732"
                  >
                    <div
                      class="_meta_d75732"
                    >
                      <div
                        class="_tags_d75732"
                      >
                        <div
                          class="_tag_d75732"
                        >
                          1 min
                        </div>
                      </div>
                    </div>
                  </div>
                </a>
              </div>
              <div
                aria-colindex="1"
                class="_cell_19b484"
                id="layout_grid_0-1"
                role="gridcell"
                style="width: 33%;"
              >
                <a
                  class="_card_d75732"
                  data-testid="Big Buck Bunny"
                  href="/m/aaabbbcd/big-buck-bunny?r=bbbdddff"
                  role="button"
                  tabindex="-1"
                >
                  <div
                    class="_titleContainer_d75732"
                  >
                    <h3
                      class="_title_d75732"
                    >
                      Big Buck Bunny
                    </h3>
                  </div>
                  <div
                    class="_poster_d75732 _aspect169_d75732"
                  >
                    <div
                      class="_meta_d75732"
                    >
                      <div
                        class="_tags_d75732"
                      >
                        <div
                          class="_tag_d75732"
                        >
                          100 min
                        </div>
                      </div>
                    </div>
                  </div>
                </a>
              </div>
              <div
                aria-colindex="2"
                class="_cell_19b484"
                id="layout_grid_0-2"
                role="gridcell"
                style="width: 33%;"
              >
                <a
                  class="_card_d75732"
                  data-testid="My last favorite"
                  href="/m/ggaaccvv/my-last-favorite?r=bbbbbbbb"
                  role="button"
                  tabindex="-1"
                >
                  <div
                    class="_titleContainer_d75732"
                  >
                    <h3
                      class="_title_d75732"
                    >
                      My last favorite
                    </h3>
                  </div>
                  <div
                    class="_poster_d75732 _aspect169_d75732"
                  >
                    <div
                      class="_meta_d75732"
                    >
                      <div
                        class="_tags_d75732"
                      >
                        <div
                          class="_tag_d75732"
                        >
                          11 min
                        </div>
                      </div>
                    </div>
                  </div>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`User Component tests > Payments Page 1`] = `
<div>
  <div
    class="_user_e9c5ca"
  >
    <nav
      class="_leftColumn_e9c5ca"
    >
      <div
        class="_panel_e9c5ca"
      >
        <ul>
          <li>
            <a
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              href="/u/my-account"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12,19.2C9.5,19.2 7.29,17.92 6,16C6.03,14 10,12.9 12,12.9C14,12.9 17.97,14 18,16C16.71,17.92 14.5,19.2 12,19.2M12,5A3,3 0 0,1 15,8A3,3 0 0,1 12,11A3,3 0 0,1 9,8A3,3 0 0,1 12,5M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12C22,6.47 17.5,2 12,2Z"
                  />
                </svg>
              </div>
              <span>
                nav.account
              </span>
            </a>
          </li>
          <li>
            <a
              aria-current="page"
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296 _active_f8f296 "
              href="/u/payments"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5,3C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V16.72C21.59,16.37 22,15.74 22,15V9C22,8.26 21.59,7.63 21,7.28V5A2,2 0 0,0 19,3H5M5,5H19V7H13A2,2 0 0,0 11,9V15A2,2 0 0,0 13,17H19V19H5V5M13,9H20V15H13V9M16,10.5A1.5,1.5 0 0,0 14.5,12A1.5,1.5 0 0,0 16,13.5A1.5,1.5 0 0,0 17.5,12A1.5,1.5 0 0,0 16,10.5Z"
                  />
                </svg>
              </div>
              <span>
                nav.payments
              </span>
            </a>
          </li>
          <li
            class="_logoutLi_e9c5ca"
          >
            <button
              class="_button_f8f296 _button_e9c5ca _default_f8f296 _text_f8f296"
              type="button"
            >
              <div
                class="_startIcon_f8f296"
              >
                <svg
                  aria-hidden="true"
                  class="_icon_585b29"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19,3H5C3.89,3 3,3.89 3,5V9H5V5H19V19H5V15H3V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V5C21,3.89 20.1,3 19,3M10.08,15.58L11.5,17L16.5,12L11.5,7L10.08,8.41L12.67,11H3V13H12.67L10.08,15.58Z"
                  />
                </svg>
              </div>
              <span>
                nav.logout
              </span>
            </button>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="_mainColumn_e9c5ca"
    >
      <h1
        class="hideUntilFocus"
      >
        nav.payments
      </h1>
      <section
        aria-labelledby="subscription-details_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="subscription-details_1235"
          >
            user:payment.subscription_details
          </h2>
        </div>
        <div
          class="_infoBox_c57ff5"
        >
          <p>
            <strong>
              user:payment.annual_subscription
            </strong>
             
            <br />
            user:payment.next_billing_date_on
          </p>
          <p
            class="_price_c57ff5"
          >
            <strong>
              $199.99
            </strong>
            <small>
              /
              account:periods.year
            </small>
          </p>
        </div>
        <button
          aria-disabled="false"
          class="_button_f8f296 _upgradeSubscription_c57ff5 _primary_f8f296 _outlined_f8f296"
          data-testid="change-subscription-button"
          type="button"
        >
          <span>
            user:payment.change_subscription
          </span>
        </button>
      </section>
      <section
        aria-labelledby="payment-method_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="payment-method_1235"
          >
            user:payment.payment_method
          </h2>
        </div>
        <div>
          <div
            class="_textField_e16c1b"
          >
            <label
              class="_label_e16c1b"
              for="text-field_1235_cardnumber"
            >
              user:payment.card_number
            </label>
            <input
              aria-describedby="helper_text_1235_cardnumber"
              aria-invalid="false"
              aria-label="user:payment.card_number_hidden"
              autocomplete="cc-number"
              class="_input_e16c1b"
              id="text-field_1235_cardnumber"
              name="cardNumber"
              readonly=""
              type="text"
              value="•••• •••• •••• 9888"
            />
          </div>
          <div
            class="_cardDetails_c57ff5"
          >
            <div
              class="_textField_e16c1b"
            >
              <label
                class="_label_e16c1b"
                for="text-field_1235_cardexpiry"
              >
                user:payment.expiry_date
              </label>
              <input
                aria-describedby="helper_text_1235_cardexpiry"
                aria-invalid="false"
                autocomplete="cc-exp"
                class="_input_e16c1b"
                id="text-field_1235_cardexpiry"
                name="cardExpiry"
                readonly=""
                type="text"
                value="03/30"
              />
            </div>
            <div
              class="_textField_e16c1b"
            >
              <label
                class="_label_e16c1b"
                for="text-field_1235_cardsecuritycode"
              >
                user:payment.security_code
              </label>
              <input
                aria-describedby="helper_text_1235_cardsecuritycode"
                aria-invalid="false"
                aria-label="user:payment.security_code_hidden"
                autocomplete="cc-csc"
                class="_input_e16c1b"
                id="text-field_1235_cardsecuritycode"
                name="cardSecurityCode"
                readonly=""
                type="text"
                value="******"
              />
            </div>
          </div>
          <button
            class="_button_f8f296 _default_f8f296 _outlined_f8f296"
            type="button"
          >
            <span>
              account:payment.edit_card
            </span>
          </button>
        </div>
      </section>
      <section
        aria-labelledby="billing-history_1235"
        class="_panel_e9c5ca"
      >
        <div
          class="_panelHeader_e9c5ca"
        >
          <h2
            id="billing-history_1235"
          >
            user:payment.billing_history
          </h2>
        </div>
        <div
          class="_infoBox_c57ff5"
        >
          <p
            class="_transactionItem_c57ff5"
          >
            <strong>
              July 16, 2022
            </strong>
            <span>
              Your Cool Subscription
            </span>
            <span>
              11232
            </span>
          </p>
          <div
            class="_transactionDetails_c57ff5"
          >
            <div
              class="_transationPrice_c57ff5"
            >
              user:payment.price_paid_with
            </div>
          </div>
        </div>
        <div
          class="_infoBox_c57ff5"
        >
          <p
            class="_transactionItem_c57ff5"
          >
            <strong>
              November 9, 2022
            </strong>
            <span>
              Your Cool Subscription
            </span>
            <span>
              11234
            </span>
          </p>
          <div
            class="_transactionDetails_c57ff5"
          >
            <div
              class="_transationPrice_c57ff5"
            >
              user:payment.price_paid_with
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;
